tests/state_sampler_test.cc

```diff
@@ -0,0 +1,136 @@
+#include <complex>
+#include <memory>
+
+#include "gtest/gtest.h"
+
+#include "../lib/gates_appl.h"
+#include "../lib/gates_def.h"
+#include "../lib/parfor.h"
+#include "../lib/simulator_avx.h"
+#include "../lib/state_sampler.h"
+#include "../lib/statespace.h"
+
+namespace qsim {
+
+TEST(SampleStatesTest, ZeroState) {
+  unsigned num_qubits = 2;
+  unsigned num_threads = 1;
+
+  using Simulator = SimulatorAVX<ParallelFor>;
+  using StateSpace = typename Simulator::StateSpace;
+  using StateSampler = StateSampler<StateSpace>;
+
+  StateSpace state_space(num_qubits, num_threads);
+  Simulator simulator(num_qubits, num_threads);
+  StateSampler sampler;
+
+  auto state = state_space.CreateState();
+  state_space.SetStateZero(state);
+
+  unsigned num_samples = 100000;
+  unsigned seed = 0;
+
+  // Prior to applying gates, only the zero state should be observed.
+  std::vector<float> expected_prob = {1, 0, 0, 0};
+  std::vector<float> counts = {0, 0, 0, 0};
+  std::vector<uint64_t> measurements;
+  sampler.SampleState(state_space, state, num_samples, seed, &measurements);
+  ASSERT_EQ(measurements.size(), num_samples);
+  for (unsigned i = 0; i < measurements.size(); ++i) {
+    int val = measurements[i];
+    counts[val] += 1.0;
+  }
+  for (unsigned i = 0; i < counts.size(); ++i) {
+    EXPECT_NEAR(counts[i] / num_samples, expected_prob[i], 1e-2);
+  }
```

Comment:
I don't think you need such an elaborate mechanism here since you have no chance of sampling anything other than 1. Maybe you could just reduce this to:

```c++
for (int i =0;i<measurements.size();i++) {
    EXPECT_EQ(val, 1);
}
```
And reduce the samples down from 100,000 ?

Reply:
Good point. `num_samples` at 100k doesn't really increase test runtime, but I dropped it to 1k just in case.